web/src/lib/utils/tree-utils.ts

```diff
@@ -1,21 +1,146 @@
-export interface RecursiveObject {
-  [key: string]: RecursiveObject;
-}
+/* eslint-disable @typescript-eslint/no-this-alias */
+/* eslint-disable unicorn/no-this-assignment */
+/* eslint-disable unicorn/prefer-at */
+import type { TagResponseDto } from '@immich/sdk';
+
+export class TreeNode extends Map<string, TreeNode> {
+  value: string;
+  path: string;
+  parent: TreeNode | null;
+  hasAssets: boolean;
+  id: string | undefined;
+  color: string | undefined;
+  private _parents: TreeNode[] | undefined;
+  private _children: TreeNode[] | undefined;
 
-export const normalizeTreePath = (path: string) => path.replace(/^\//, '').replace(/\/$/, '');
+  private constructor(value: string, path: string, parent: TreeNode | null) {
+    super();
+    this.value = value;
+    this.parent = parent;
+    this.path = path;
+    this.hasAssets = false;
+  }
 
-export function buildTree(paths: string[]) {
-  const root: RecursiveObject = {};
+  static fromPaths(paths: string[]) {
+    const root = new TreeNode('', '', null);
+    for (const path of paths) {
+      const current = root.add(path);
+      current.hasAssets = true;
+    }
+    return root;
+  }
 
-  for (const path of paths) {
-    const parts = path.split('/');
-    let current = root;
+  static fromTags(tags: TagResponseDto[]) {
+    const root = new TreeNode('', '', null);
+    for (const tag of tags) {
+      const current = root.add(tag.value);
+      current.hasAssets = true;
+      current.id = tag.id;
+      current.color = tag.color;
+    }
+    return root;
+  }
+
+  traverse(path: string) {
+    const parts = getPathParts(path);
+    let current: TreeNode = this;
+    let curPart = null;
     for (const part of parts) {
-      if (!current[part]) {
-        current[part] = {};
+      // segments common to all subtrees can be collapsed together
+      curPart = curPart === null ? part : joinPaths(curPart, part);
+      const next = current.get(curPart);
+      if (next) {
+        current = next;
+        curPart = null;
       }
-      current = current[part];
     }
+    return current;
+  }
+
+  collapse() {
+    if (this.size === 1 && !this.hasAssets) {
+      const child = this.values().next().value!;
+      child.value = joinPaths(this.value, child.value);
+      child.parent = this.parent;
+      if (this.parent !== null) {
+        this.parent.delete(this.value);
+        this.parent.set(child.value, child);
+      }
+    }
+
+    for (const child of this.values()) {
+      child.collapse();
+    }
+  }
+
+  private add(path: string) {
+    let current: TreeNode = this;
+    for (const part of getPathParts(path)) {
+      let next = current.get(part);
+      if (next === undefined) {
+        next = new TreeNode(part, joinPaths(current.path, part), current);
```

Comment:
You'd then do something like `next = new TreeNode(part, joinPaths(current.path, part), add(remainingPath))`